src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,283 @@
+  }
+}
+
+// The rows of vector and matrix will be reordered using permutationMatrix
+void reorder(const Eigen::MatrixXd& permutationMatrix,
+  const Eigen::VectorXd& inVector,
+  const Eigen::MatrixXd& inMatrix,
+  Eigen::VectorXd* outVector,
+  Eigen::MatrixXd* outMatrix)
+{
+  *outVector = permutationMatrix*inVector;
+  *outMatrix = permutationMatrix*inMatrix;
+}
+
+// Convert the mapping to a permutation matrix for matrix operation
+void convertToPermutationMatrix(const std::map<size_t, size_t>& map,
+  Eigen::MatrixXd* permutationMatrix)
+{
+  *permutationMatrix = Eigen::MatrixXd::Zero(map.size(), map.size());
+  for(auto it = map.begin(); it != map.end(); ++it)
+  {
+    (*permutationMatrix)(it->second, it->first) = 1.0;
+  }
+}
+
+} // namespace
+
+//=============================================================================
+std::unique_ptr<SplineTrajectory> convertJointTrajectory(
+  const std::shared_ptr<MetaSkeletonStateSpace>& space,
+  const trajectory_msgs::JointTrajectory& jointTrajectory)
+{
+  if (!space)
+    throw std::invalid_argument{"StateSpace must be non-null."};
+
+  const auto numControlledJoints = space->getNumSubspaces();
+  if (jointTrajectory.joint_names.size() != numControlledJoints)
+  {
+    std::stringstream message;
+    message << "Incorrect number of joints: expected "
+        << numControlledJoints << ", got "
+        << jointTrajectory.joint_names.size() << ".";
+    throw std::invalid_argument{message.str()};
+  }
+
+  // Check that all joints are single DOF.
+  for (size_t i = 0; i < space->getNumSubspaces(); ++i)
+  {
+    auto n = space->getJointSpace(i)->getJoint()->getNumDofs();
+    if (n != 1)
+    {
+      std::stringstream message;
+      message << "Expected 1 dof. Joint " << i << " has " << n << " dofs.";
+      throw std::invalid_argument{message.str()};
+    }
+  }
+
+  if (jointTrajectory.points.size() < 2)
+  {
+    throw std::invalid_argument{
+      "Trajectory must contain two or more waypoints."};
+  }
+
+  // Map joint indices between jointTrajectory and space subspaces.
+  std::map<size_t, size_t> rosJointToMSSSJoint;
+
+  for (size_t i = 0; i < jointTrajectory.joint_names.size(); ++i)
+  {
+    std::string joint_name = jointTrajectory.joint_names[i];
+    bool found_match = false;
+
+    for (size_t j = 0; j < space->getNumSubspaces(); ++j)
+    {
+      auto joint = space->getJointSpace(j)->getJoint();
+      if (joint_name == joint->getName())
+      {
+        // Check that the mapping is unique.
+        auto search = rosJointToMSSSJoint.find(i);
+        if (search != rosJointToMSSSJoint.end())
+        {
+          std::stringstream message;
+          message << "Both subspace[" << search->second << "] and ["
+            << j << "] map to joint " << i << " in jointTrajectory.";
+          throw std::invalid_argument{message.str()};
+        }
+        rosJointToMSSSJoint.emplace(std::make_pair(i, j));
+        found_match = true;
+        break;
+      }
```

Comment:
Encapsulate this in a helper function. Putting this search inline means that we can't easily improve it's efficiency later (e.g. using `unordered_map`) and makes it harder to understand this function.

I suggest writing a helper function to get a `Joint` by name from a `MetaSkeleton`. If the name is ambiguous, which can occur in a `MetaSkeleton` that contain `Joint`s from multiple `Skeleton`s, `throw` an exception. Then, this code simply becomes:

        auto metaSkeleton = space->getMetaSkeleton();
        auto joint = getJointByName(*metaSkeleton, jointName);
        auto index = metaSkeleton->getIndex(joint);